xcm/xcm-builder/src/pay.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! `PayOverXcm` struct for paying through XCM and getting the status back
+
+use frame_support::traits::{
+	tokens::{Pay, PaymentStatus},
+	Get,
+};
+use sp_std::{marker::PhantomData, vec};
+use xcm::prelude::*;
+use xcm_executor::traits::{Convert, QueryResponseStatus, XcmQueryHandler};
+
+/// Implementation of the `frame_support_traits::tokens::Pay` trait, to allow
+/// for generic payments of a given `AssetKind` and `Balance` from an implied origin, to a
+/// beneficiary via XCM, relying on the XCM `TransferAsset` instruction.
+///
+/// `PayOverXcm::pay` is asynchronous, and returns a `QueryId` which can then be used in
+/// `check_payment` to check the status of the XCM transaction.
+///
+pub struct PayOverXcm<
+	DestChain,
+	Router,
+	Querier,
+	BlockNumber,
+	Timeout,
+	AccountId,
+	AccountIdConverter,
+>(PhantomData<(DestChain, Router, Querier, BlockNumber, Timeout, AccountId, AccountIdConverter)>);
+impl<
+		DestChain: Get<xcm::latest::MultiLocation>,
+		Router: SendXcm,
+		Querier: XcmQueryHandler,
+		BlockNumber,
+		Timeout: Get<Querier::BlockNumber>,
+		AccountId: Clone,
+		AccountIdConverter: Convert<MultiLocation, AccountId>,
+	> Pay
+	for PayOverXcm<DestChain, Router, Querier, BlockNumber, Timeout, AccountId, AccountIdConverter>
+{
+	type Beneficiary = AccountId;
+	type AssetKind = xcm::latest::AssetId;
+	type Balance = u128;
+	type Id = Querier::QueryId;
+
+	fn pay(
+		who: &Self::Beneficiary,
+		asset_kind: Self::AssetKind,
+		amount: Self::Balance,
+	) -> Result<Self::Id, ()> {
+		let beneficiary = AccountIdConverter::reverse(who.clone()).map_err(|_| ())?;
+		let mut message = Xcm(vec![
+			UnpaidExecution { weight_limit: Unlimited, check_origin: None },
```

Comment:
A way to transfer from the treasury account, is to add here DescendOrigin instruction, we set it to multilocation of here/treasury_account_id

Reply:
`DescendOrigin` goes down from the origin of the message. This message is interpreted by statemint, what's the origin going to be? How does this instruction work in this case?

Reply:
On statemint XCM program will start the execution as (parent: 0, here), from this root multi-location we descend to account id.
This how I get it from the code base.

Reply:
If the Collectives chain sends this message to Statemint (as I expect to be the case), then this XCM will begin with the Origin register set to `( parents: 1, [Parachain(Collectives)] )`. Descending from there into an account will likely not have the desired effect (it will depend on the exact impl of `SovereignAccountOf` in the asset transactor).

Basically, there's no point in having `sender_account` here unless we have configured `SovereignAccountOf` to alias accounts from this chain. A possibly better way of doing things would be to use `DescendOrigin` with a more specialised junction to denote the body which owns/controls the pot, perhaps either `Junction::PalletInstance` or `Junction::Plurality`. My suggestion above for a `DescendJunction` type param instead of `SenderAccount` leaves it open.